src/scportrait/pipeline/extraction.py

```diff
@@ -524,206 +509,145 @@ def _extract_classes(self, px_center, arg, return_failed_ids=False):
-
-            self.save_index_to_remove.append(save_index)
-            self._save_failed_cell_info(
-                save_index,
-                nucleus_id,
-                image_index,
-                label_info,
-            )
+                fig, axs = plt.subplots(1, stack.shape[0], figsize=(2 * stack.shape[0], 2))
+                for i, img in enumerate(stack):
+                    axs[i].imshow(img, vmin=0, vmax=1)
+                    axs[i].axis("off")
+                fig.tight_layout()
+                fig.show()
 
-            if return_failed_ids:
-                return [save_index]
-            else:
-                return None
+        if return_results:
+            return save_index, stack, cell_id
+        else:
+            self._single_cell_data[save_index] = stack
+            self._single_cell_index[save_index] = [save_index, cell_id]
+            return None
 
-    def _extract_classes_multi(self, px_centers, arg_list):
+    def _extract_classes_multi(
+        self,
+        arg_list,
+    ):
         self._setup_normalization()
 
         self.seg_masks = mmap_array_from_path(self.path_seg_masks)
         self.image_data = mmap_array_from_path(self.path_image_data)
 
-        self._tmp_single_cell_index = mmap_array_from_path(self._tmp_single_cell_index_path)
-        self._tmp_single_cell_data = mmap_array_from_path(self._tmp_single_cell_data_path)
-
-        results = []
-        for arg in arg_list:
-            x = self._extract_classes(px_centers, arg, return_failed_ids=True)
-            results.append(x)
-
-        return flatten(results)
-
-    def _transfer_tempmmap_to_hdf5(self):
-        self.log("Transferring results to final HDF5 data container.")
-
-        # reconnect to memory mapped temp arrays
-        _tmp_single_cell_index = mmap_array_from_path(self._tmp_single_cell_index_path)
-        _tmp_single_cell_data = mmap_array_from_path(self._tmp_single_cell_data_path)
-
-        self.log(f"number of cells too close to image edges to extract: {len(self.save_index_to_remove)}")
+        # get processing results
+        results = [self._extract_classes(arg, return_results=True) for arg in arg_list]
 
-        # generate final index of all of the rows that we wish to keep out of the original array
-        keep_index = np.setdiff1d(np.arange(_tmp_single_cell_index.shape[0]), self.save_index_to_remove)
+        return results
 
-        # get cell_ids of the cells that were successfully extracted
-        _, cell_ids = _tmp_single_cell_index[keep_index].T
-        _, cell_ids_removed = _tmp_single_cell_index[self.save_index_to_remove].T
+    def _write_to_hdf5(self, results, hdf5_lock):
+        """Function for writing results to HDF5 file in a thread-safe manner."""
+        with hdf5_lock:
+            with h5py.File(self.output_path, "a") as hf:
+                self._single_cell_data = hf["single_cell_data"]
```

Comment:
```suggestion
                self._single_cell_data_container = hf["single_cell_data"]
```
or somethign to suggest this isn't in-memory